### PR TITLE
circuit: pubdata optimization

### PR DESCRIPTION
**Pull request by lightning-li**
### Description

The pubdata need to be posted on L1 chain, and now pubdata size of each transaction is 192 bytes which need to cost lots of gas. We should compact pubdata size as much as possible. After this optimization, the pubdata size reduces from 192 bytes to 121 bytes.  

### Changes

Notable changes:
* Arrange pubdata as bits rather than Variable
* Remove `nftL1Address` and `nftL1TokenId` from related tx's pubdata
* Remove `gasAccountIndex` from all kinds of tx's pubdata
* Changes `accountName` length from 32 to 20 bytes

### CODE REVIEW

File: circuit/types/hint.go [hunk 1]
```diff
@@ -8,13 +8,41 @@ import (
...
+//	result := outputs[0]
+//	result.SetBytes(hashVal[:])
+//	return nil
+//}
+
 func Keccak256(_ ecc.ID, inputs []*big.Int, outputs []*big.Int) error {
 	var buf bytes.Buffer
-	for i := 0; i < len(inputs); i++ {
+	for i := 0; i < 4; i++ {
 		buf.Write(inputs[i].FillBytes(make([]byte, 32)))
```
Comment (ruslangm) [inline circuit/types/hint.go:27]: shouldn't we add simple explanation comment for a new hashing scheme for the block commitment? like:
```go
// process blockNumber, createdAt, oldStateRoot, newStateRoot
for i := 0; i < 4; i++ {
	buf.Write(inputs[i].FillBytes(make([]byte, 32)))
}

// process pubData
for i := 4; i < len(inputs)-1; i = i + 8 {
	buf.Write(CovertBitsToBigInt(inputs[i : i+8]).FillBytes(make([]byte, 1)))
}

// process onChainOpsCount
buf.Write(inputs[len(inputs)-1].FillBytes(make([]byte, 32)))

```

File: circuit/types/pubdata_helper.go [hunk 2]
```diff
@@ -17,355 +17,498 @@
 
 package types
 
-func CollectPubDataFromRegisterZNS(api API, txInfo RegisterZnsTxConstraints) (pubData [PubDataSizePerTx]Variable) {
+func CollectPubDataFromRegisterZNS(api API, txInfo RegisterZnsTxConstraints) (pubData [PubDataBitsSizePerTx]Variable) {
+	currentOffset := 0
 	txTypeBits := api.ToBinary(TxTypeRegisterZns, TxTypeBitsSize)
+	CopyLittleEndianSlice(pubData[currentOffset:currentOffset+TxTypeBitsSize], txTypeBits)
```
Comment (ruslangm) [inline circuit/types/pubdata_helper.go]: Does it make sense to create a separate func for repeated pieces of code and use it then for different types? smth like:
```go
func CollectPubDataFromRegisterZNS(api API, txInfo RegisterZnsTxConstraints) (pubData [PubDataBitsSizePerTx]Variable) {
        currentOffset := 0
        copyLittleEndianSliceAndShiftOffset(api, TxTypeRegisterZns, TxTypeBitsSize, &currentOffset, pubData)
        copyLittleEndianSliceAndShiftOffset(api, txInfo.AccountIndex, AccountIndexBitsSize, &currentOffset, pubData)
        ...
}

func copyLittleEndianSliceAndShiftOffset(api API, txType Variable, txTypeBitsSize int, currentOffset *int, pubData [PubDataBitsSizePerTx]Variable) {
	txTypeBits := api.ToBinary(txType, txTypeBitsSize)
	CopyLittleEndianSlice(pubData[*currentOffset:*currentOffset+txTypeBitsSize], txTypeBits)
	*currentOffset += TxTypeBitsSize
}
```

**Review by ruslangm — APPROVED** [review]

File: circuit/types/pubdata_helper.go (hunk 2 above)
Comment (lightning-li) [reply circuit/types/pubdata_helper.go]: Yeah, it's a good suggestion to make the code more clear, though it doesn't reduce the constraints counts. I will fix this as you suggested.

File: circuit/types/hint.go (hunk 1 above)
Comment (lightning-li) [reply circuit/types/hint.go:27]: I will add some comments for this change.

**Pull request merged**